pkg/envoy/xds/cache.go

```diff
@@ -105,27 +109,49 @@ func (c *Cache) tx(typeURL string, upsertedResources map[string]proto.Message, d
 		// responses in GetResources.
 		// Calling proto.Message.String is not very cheap, but we assume that
 		// the reduced churn between the clients and the server is worth it.
-		if force || !found || oldV.resource.String() != value.String() {
+		if !found || oldV.resource.String() != value.String() {
 			if found {
 				cacheLog.WithField(logfields.XDSResourceName, name).Debug("updating resource in cache")
+
+				if revertUpsertedResources == nil {
+					revertUpsertedResources = make(map[string]proto.Message, len(upsertedResources)+len(deletedNames))
+				}
+				revertUpsertedResources[name] = oldV.resource
```

Comment:
I'd call this variable `revertChangedResources`...

Reply:
This map will be passed as parameter `upsertedResources`, so I wanted to match that name.